### ISSUE TITLE
Memory leak and possible stackoverflowexception in ReadEventsInternalAsync

### ISSUE THREAD

**Issue by arnodenuijl** [EventStore/EventStore#1363]
The interaction between **ReadEventsInternalAsync** and **ReadEventsCallbackAsync** in **EventStoreCatchUpSubscription.cs** can cause memory pressure and can ultimately result in a stackoverflowexception.  (https://github.com/EventStore/EventStore/blob/release-v4.0.2/src/EventStore.ClientAPI/EventStoreCatchUpSubscription.cs#L541 & https://github.com/EventStore/EventStore/blob/release-v4.0.2/src/EventStore.ClientAPI/EventStoreCatchUpSubscription.cs#L548)

The two methods call each other which results in a big stack that takes up a decent amount of memory and, if continued long enough will result in a stackoverflowexception. See https://groups.google.com/d/msg/event-store/FrQEHpqCPhs/cR3zsB1_BQAJ for an example of a growing stack.

Easy way to reproduce is to just subscribe to a very large stream and wait. 
```
var es = EventStoreConnectionFactory.CreateConnection();
 es.SubscribeToStreamFrom("$ce-qbox", 0, CatchUpSubscriptionSettings.Default, (s, e) =>
 {

 });
```

I think there is an easy fix by replacing the mutually recursive function call with a while loop. I will make a pull request for it to show it.

**Comment by gregoryyoung** [comment]
👍

**Comment by arnodenuijl** [comment]
anything I can do for this item? I created a pull request with the changes. I've been running the client from my fork and it works great. Since we have streams with millions of events, it saves a huge amount of memory. 
I saw the pull request was out of date so i clicked the 'Update branch' button, but i'm not sure that is the way to go. Should i revert the merge and rebase and force push my changes so it's in one commit?!
